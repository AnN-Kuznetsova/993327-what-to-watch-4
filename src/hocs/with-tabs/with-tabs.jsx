import React, {PureComponent} from 'react';


export const withTabs = (Component) => {
  class WithTabs extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        activeTabId: 0,
      };

      this.renderTab = this.renderTab.bind(this);
      this.renderTabNav = this.renderTabNav.bind(this);
    }

    renderTabNav(tabNames) {
      const {activeTabId} = this.state;

      return (
        <nav className="movie-nav movie-card__nav">
          <ul className="movie-nav__list">
            {tabNames.map((tabName, index) => (
              <li key={tabName + index} className={`movie-nav__item ${index === activeTabId ? `movie-nav__item--active` : ``}`}>
                <a href="#" className="movie-nav__link">{tabName}</a>
              </li>
            ))}
          </ul>
        </nav>
      );
    }

    renderTab(tabId, TabComponent, props, key) {
      const {activeTabId} = this.state;

      if (tabId === activeTabId) {
        return (
          <TabComponent key={key} {...props} />
        );
      }

      return null;
    }

    render() {
      return <Component
        {...this.props}
        renderTabNav={this.renderTabNav}
        renderTab={this.renderTab}
      />;
    }
  }


  WithTabs.propTypes = {};


  return WithTabs;
};
