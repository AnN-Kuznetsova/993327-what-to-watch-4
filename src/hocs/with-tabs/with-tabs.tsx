import * as React from 'react';
import {Subtract} from "utility-types";


interface State {
  activeTabId: number;
}

interface InjectingProps {
  renderTabNav: (tabNames: string[]) => React.ReactNode;
  renderTab: (tabId: number, TabComponent: React.FunctionComponent, props: object, key: string) => React.ReactNode;
}

export const withTabs = (Component) => {
  type P = React.ComponentProps<typeof Component>;
  type T = Subtract<P, InjectingProps>;

  class WithTabs extends React.PureComponent<T, State> {
    constructor(props) {
      super(props);

      this.state = {
        activeTabId: 0,
      };

      this.renderTab = this.renderTab.bind(this);
      this.renderTabNav = this.renderTabNav.bind(this);
    }

    handleTabHeaderClick(tabId) {
      this.setState({
        activeTabId: tabId,
      });
    }

    renderTabNav(tabNames) {
      const {activeTabId} = this.state;

      return (
        <nav className="movie-nav movie-card__nav">
          <ul className="movie-nav__list">
            {tabNames.map((tabName, index) => (
              <li key={tabName + index} className={`movie-nav__item ${index === activeTabId ? `movie-nav__item--active` : ``}`}>
                <a
                  href="#"
                  className="movie-nav__link"
                  onClick={this.handleTabHeaderClick.bind(this, index)}
                >{tabName}</a>
              </li>
            ))}
          </ul>
        </nav>
      );
    }

    renderTab(tabId, TabComponent, props, key) {
      const {activeTabId} = this.state;

      if (tabId === activeTabId) {
        return (
          <TabComponent
            {...props}
            key={key}
          />
        );
      }

      return null;
    }

    render() {
      return <Component
        {...this.props}
        renderTabNav={this.renderTabNav}
        renderTab={this.renderTab}
      />;
    }
  }


  return WithTabs;
};
