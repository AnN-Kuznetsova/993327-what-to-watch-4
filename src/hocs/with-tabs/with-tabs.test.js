import PropTypes from "prop-types";
import React from "react";
import renderer from "react-test-renderer";

import {withTabs} from "./with-tabs";


const tabList = [
  {
    name: `One`,
    // eslint-disable-next-line react/display-name
    component: () => <div className="tab-content 1" />,
  },
  {
    name: `Two`,
    // eslint-disable-next-line react/display-name
    component: () => <div className="tab-content 2" />,
  },
  {
    name: `Three`,
    // eslint-disable-next-line react/display-name
    component: () => <div className="tab-content 3" />,
  },
];

const Component = (props) => {
  const {renderTabNav, renderTab} = props;
  const tabNames = tabList.map((tab) => tab.name);

  return (
    <div>
      {renderTabNav(tabNames)}
      {renderTab(0, tabList[0].component, ({}), `content1`)}
      {renderTab(1, tabList[1].component, ({}), `content2`)}
      {renderTab(2, tabList[2].component, ({}), `content3`)}
    </div>
  );
};

Component.propTypes = {
  renderTabNav: PropTypes.func.isRequired,
  renderTab: PropTypes.func.isRequired,
};

const ComponentWithTabs = withTabs(Component);


describe(`Render withTabs`, () => {
  it(`Should match with snapshot`, () => {
    const componentWithTabsSnapshot = renderer.create(
        <ComponentWithTabs />
    ).toJSON();

    expect(componentWithTabsSnapshot).toMatchSnapshot();
  });
});
