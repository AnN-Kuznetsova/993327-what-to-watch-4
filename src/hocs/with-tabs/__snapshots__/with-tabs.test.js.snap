// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render withTabs Should match with snapshot 1`] = `
<div>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          One
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Two
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Three
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="tab-content 1"
  />
</div>
`;
